cpp/src/arrow/dataset/dataset.cc

```diff
@@ -202,11 +202,7 @@ Result<FragmentIterator> InMemoryDataset::GetFragmentsImpl(compute::Expression)
 
   auto create_fragment =
       [schema](std::shared_ptr<RecordBatch> batch) -> Result<std::shared_ptr<Fragment>> {
-    if (!batch->schema()->Equals(schema)) {
-      return Status::TypeError("yielded batch had schema ", *batch->schema(),
-                               " which did not match InMemorySource's: ", *schema);
-    }
-
+    RETURN_NOT_OK(CheckProjectable(*schema, *batch->schema()));
```

Comment:
It _feels_ like this could be a construction-time check to avoid repeated checking except there is no way to return a Status there, unfortunately. (Not a big deal, though.)

Reply:
I thought about that, but would have to change this to a `::Make()` method and didn't want to go that far here.